File: javascript/immutability-java-vs-js/immutability.java
public class Immutability {
    public static void main(String[] args) {
        List<String> mutable = new ArrayList<>(List.of("alfa", "beta"));
        mutable.add("gamma");
        System.out.println(mutable);

        List<String> seedData = List.of("alpha", "beta", "gamma");
        try {
            seedData.add("delta");
        } catch (UnsupportedOperationException e) {
            System.out.println("add() threw UnsupportedOperationException, as expected");
        }

        List<String> backing = new ArrayList<>(List.of("x", "y"));
        List<String> view = Collections.unmodifiableList(backing);

        backing.add("z");
        System.out.println(view); // [x, y, z]
    }
}

File: javascript/immutability-java-vs-js/immutability.js
"use strict";

const plain = ["alpha", "beta", "gamma"];
plain.push("delta");
console.log(plain);

const seedData = Object.freeze(["alpha", "beta", "gamma"]);
try {
  seedData.push("delta");
} catch (e) {
  console.log(e.constructor.name, "-", e.message);
  // TypeError - Cannot add property 3, object is not extensible
}

const shallow = Object.freeze([
  { id: 1, name: "alpha" },
  { id: 2, name: "beta" },
]);

shallow[0].name = "MUTATED";
console.log(JSON.stringify(shallow));

console.log(Object.isFrozen(shallow)); // true
console.log(Object.isFrozen(shallow[0])); // false